### ISSUE TITLE
Function registration and reaction_func

### ISSUE THREAD

**Issue by ChrisRackauckas** [SciML/Catalyst.jl#248]
`@reaction_func` can probably just be removed and instead it can rely on direct tracing. An example of how `@reaction_func` can lead to issues is:

```julia
using Catalyst, DiffEqFlux, ModelingToolkit

Q_val(S, I, R, p) = FastChain(FastDense(3,10,relu),FastDense(10,1))([S, I, R], p)
@reaction_func NN(S,I,R,p) = Q_val(S, I, R, p)
@register Q_val(S, I, R, p)

rn = @reaction_network begin
    β, S + I --> 2I
    γ, I --> R
    NN(S, I, R, p3n), I --> Q
    δ, Q --> R
end β γ δ p3n
```

In this case, we want `Q_val` to be captured as a node in the compute graph, but the `reaction_func` mechanism always traces and it fails at `FastChain`. Instead of having a separate mechanism, Catalyst can instead just build the variables right away and trace via substitution.

**Comment by isaacsas** [comment]
Yeah, this was something I was wondering about during the rewrite and then lost track...  We should update for this.

**Comment by TorkelE** [comment]
Changing how functions work sounds sensible. I have to admit, I only vaguely know how the current one, as well an updated one, would works.

**Comment by isaacsas** [comment]
This seems like another case where we need a function that takes an expression and creates a MT operation, like what @ChrisRackauckas sent me on Slack (though I haven't played with that yet). Then we could restructure the macro to build the variables from the parsed symbols, and create `Operation`s corresponding to the rate.

**Comment by isaacsas** [comment]
If we just remove `reaction_funct`, the issue that arises is that MT registered function symbols are not recognized. We can see what is going on using `@macroexpand`:
```julia
@register Base.sign(x)
k2(x) = 0.5*( sign(x - 0.5) + 1 )
@register k2(x)
@macroexpand rn =@reaction_network begin
             k1, A --> B
             k2(A), B --> X
       end k1
```
gives
```julia
:(rn = begin
          begin
              var"#156#t" = (Num)((Sym){ModelingToolkit.Parameter{Real}}(:t))
              var"#157#k1" = (Num)((Sym){ModelingToolkit.Parameter{Real}}(:k1))
              (var"#156#t", var"#157#k1") = (var"#156#t", var"#157#k1")
          end
          begin
              var"#158#A" = (Num)(((Sym){(SymbolicUtils.FnType){Catalyst.NTuple{1, Catalyst.Any}, Real}}(:A))((ModelingToolkit.value)(var"#156#t")))
              var"#159#B" = (Num)(((Sym){(SymbolicUtils.FnType){Catalyst.NTuple{1, Catalyst.Any}, Real}}(:B))((ModelingToolkit.value)(var"#156#t")))
              var"#160#X" = (Num)(((Sym){(SymbolicUtils.FnType){Catalyst.NTuple{1, Catalyst.Any}, Real}}(:X))((ModelingToolkit.value)(var"#156#t")))
              (var"#158#A", var"#159#B", var"#160#X") = (var"#158#A", var"#159#B", var"#160#X")
          end
          Catalyst.ReactionSystem([Catalyst.Reaction(var"#157#k1", [var"#158#A"], [var"#159#B"], [1], [1], only_use_rate = false), Catalyst.Reaction(Catalyst.k2(var"#158#A"), [var"#159#B"], [var"#160#X"], [1], [1], only_use_rate = false)], var"#156#t", [var"#158#A", var"#159#B", var"#160#X"], [var"#157#k1"])
      end)
```
So the issue is just that we get `Catalyst.k2(var"#158#A")` instead of `k2(var"#158#A")`.  I'm not super comfortable with macros and escaping, but I guess we need to identify any non-parameter / species symbols and escape them within the expression?

**Comment by isaacsas** [comment]
This seems to go through with https://github.com/SciML/Catalyst.jl/pull/286
```julia
using Catalyst, DiffEqFlux, ModelingToolkit
Q_val(S, I, R, p) = FastChain(FastDense(3,10,relu),FastDense(10,1))([S, I, R], p)
@register Q_val(S,I,R,p)
rn = @reaction_network begin
           β, S + I --> 2I
           γ, I --> R
           Q_val(S, I, R, p3n), I --> Q
           δ, Q --> R
       end β γ δ p3n
```
I'm going to close once that PR is merged, but we should next test if conversion to concrete models works properly with such coefficients, and get that fixed in MT if there are issues there.

**Comment by isaacsas** [comment]
Closed by https://github.com/SciML/Catalyst.jl/pull/286.

Let's make a new issue if followup testing shows conversion isn't working from the generated `ReactionSystem`.

**Comment by ChrisRackauckas** [comment]
```julia
using OrdinaryDiffEq, Catalyst, DiffEqFlux, ModelingToolkit

NN(S, I, R, p) = FastChain(FastDense(3,10,tanh),FastDense(10,1))([S, I, R], p)
@register NN(S, I, R, p)

rn = @reaction_network begin
    β, S + I --> 2I
    γ, I --> R
    NN(S, I, R, p3n), I --> Q
    δ, Q --> R
end β γ δ p3n

_p3n = Float64.(initial_params(FastChain(FastDense(3,10,tanh),FastDense(10,1))))
@parameters β γ δ p3n
p     = [β=>1.0,γ=>1.0,δ=>1.0,p3n=>_p3n]
tspan = (0.0,250.0)
u0    = [999.0,0.0,1.0,0.0]
op    = ODEProblem(rn, u0, tspan, p)
sol   = solve(op,Tsit5())
```

Indeed it's good on the catalyst side. More is needed in MTK but that's a separate issue.

**Comment by isaacsas** [comment]
Given what a simple fix this ended up being, sorry it took so long! Unfortunately the past semester was just miserable for finding time to do much coding (or research actually...).

**Comment by ChrisRackauckas** [comment]
For reference, https://github.com/SciML/ModelingToolkit.jl/pull/702 works!

```julia
using OrdinaryDiffEq, Catalyst, DiffEqFlux, ModelingToolkit

NN(S, I, R, p) = FastChain(FastDense(3,10,tanh),FastDense(10,1))([S, I, R], p)[1]
@register NN(S, I, R, p)

rn = @reaction_network begin
    β, S + I --> 2I
    γ, I --> R
    NN(S, I, R, p3n), I --> Q
    δ, Q --> R
end β γ δ p3n

_p3n = Float64.(initial_params(FastChain(FastDense(3,10,tanh),FastDense(10,1))))
@parameters β γ δ p3n
p     = [β=>1.0,γ=>1.0,δ=>1.0,p3n=>_p3n]
tspan = (0.0,250.0)
u0    = [999.0,0.0,1.0,0.0]
op    = ODEProblem(rn, u0, tspan, p)
sol   = solve(op,Tsit5())
```

**Comment by isaacsas** [comment]
Awesome, we should figure out a tutorial we can use this in to show it off!
